### DISCUSSION TITLE
[Hosts] 为什么Win上不用U8 with BOM

### DISCUSSION THREAD

**Discussion by imba-tjd** [BeyondDimension/SteamTools#429] [General]
RT，讨论。

**Comment by AigioL** [comment]
可能较早版本的 Windows 上 UTF-8 with BOM 会使其失效。
参考资料：
https://www.cnblogs.com/joshua317/articles/4180191.html
https://www.cnblogs.com/heiing/archive/2012/02/17/2355429.html
https://www.cnblogs.com/waw/p/9317462.html

而且 Windows 上有一部分其他软件是以 **当前活动页编码(ANSICodePage)** 操作 hosts 文件的。
https://github.com/SteamTools-Team/SteamTools/issues/362

而且之前 V1 版本由于使用的 .NET Framework 默认编码是 **当前活动页编码(ANSICodePage)**

现在是有一个设置可以更改编码
https://github.com/SteamTools-Team/SteamTools/blob/2.4.11/src/ST.Client.Desktop/Services/Implementation/HostsFileServiceImpl.cs#L484

但是没有做 UI

**Comment by imba-tjd** [comment]
我同意不能用U8，但U8 with BOM我测试7和10是可以的。

阁下给的前三条参考资料并没有说U8 with BOM不行，而是只说了要用ANSI，因此有极大概率他们没测试U8 with BOM。

如果确定win能用U8 with BOM，在我看来，如果要尽量现代化，应设计成读允许ANSI和U8 with BOM，写只为U8 with BOM。至于其它只能读ANSI的hosts软件，那是它们的问题，won't fix。

**Comment by AigioL** [comment]
读取应该检测 BOM 使用 UTF8 with BOM 或 ANSICodePage
写入的话，还是做个设置项吧，这样也能手动改设置兼容。 @rmbadmin
